### PR TITLE
feat: to_file for assets and emojis

### PR DESCRIPTION
**Pull request by DenverCoder1**
## Summary

* Adds `to_file` to all objects that use `AssetMixin` (`Asset`, `Emoji`, `PartialEmoji`, `Sticker`)

This feature was lightly discussed [on Discord](https://canary.discord.com/channels/336642139381301249/603069307286454290/958918196021313596). The idea comes from VJ#5945.

## Example code

```py
@bot.command()
async def avatar_file(ctx: commands.Context[commands.Bot]):
    file = await ctx.author.display_avatar.to_file()
    await ctx.send(file=file)

@bot.command()
async def emoji_file(ctx: commands.Context[commands.Bot], emoji: discord.Emoji):
    file = await emoji.to_file()
    await ctx.send(file=file)

@bot.command()
async def partial_emoji_file(ctx: commands.Context[commands.Bot], partial_emoji: discord.PartialEmoji):
    file = await partial_emoji.to_file()
    await ctx.send(file=file)

@bot.command()
async def sticker_file(ctx: commands.Context[commands.Bot], message_id: int):
    msg = await ctx.channel.fetch_message(message_id)
    file = await msg.stickers[0].to_file()
    await ctx.send(file=file)
```

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/asset.py [hunk 1]
```diff
@@ -124,6 +125,39 @@ async def save(self, fp: Union[str, bytes, os.PathLike[Any], io.BufferedIOBase],
...
+
+        Returns
+        -------
+        :class:`File`
+            The asset as a file suitable for sending.
+        """
+
+        data = await self.read()
+        url = yarl.URL(self.url)
+        filename = url.path.split('/')[-1]
```
Comment (Rapptz) [inline discord/asset.py:158]:
```suggestion
        _, _, filename = url.path.rpartition('/')
```

**Review by Rapptz — COMMENTED** [review]
This looks fine, just a small nit.

File: discord/asset.py [hunk 2]
```diff
@@ -124,6 +125,39 @@ async def save(self, fp: Union[str, bytes, os.PathLike[Any], io.BufferedIOBase],
...
+
+        Parameters
+        -----------
+        spoiler: :class:`bool`
+            Whether the file is a spoiler.
+
+        Raises
+        ------
+        HTTPException
+            Downloading the asset failed.
```
Comment (Rapptz) [inline discord/asset.py:144]: Sorry, I forgot one more thing.
```suggestion
        DiscordException
            The asset does not have an associated state.
        HTTPException
            Downloading the asset failed.
```

Comment (DenverCoder1) [reply discord/asset.py:144]: Also maybe TypeError if it's a sticker of type "lottie"

Comment (Rapptz) [reply discord/asset.py:144]: Yup that'd be another case of error.

**Review by Rapptz — APPROVED** [review]
Thanks.

**Pull request merged**